packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -38,27 +107,210 @@ contract FundingRateStore is FundingRateStoreInterface, Testable {
+            msg.sender
+        );
+
+        // Delete pending proposal and copy into dispute records.
+        fundingRateRecord.proposal.disputer = msg.sender;
+        fundingRateDisputes[identifier][fundingRateRecord.proposal.time] = fundingRateRecord;
+        delete fundingRateRecords[identifier].proposal;
+    }
+
+    /**
+     * @notice Settle a disputed funding rate. The winner of the dispute, either the disputer or the proposer,
+     * will receive a reward plus their final fee bond. This method will also overwrite the current funding rate
+     * with the resolved funding rate returned by the Oracle. Pending funding rates are unaffected by this method.
+     * @dev This will revert if there is no price available for the disputed funding rate.
+     * @param identifier Identifier to settle disputed funding rate for.
+     * @param proposalTime Proposal time at which the disputed funding rate was proposed.
+     */
+    function settleDispute(bytes32 identifier, uint256 proposalTime) external nonReentrant() {
+        FundingRateRecord storage fundingRateDispute = _getFundingRateDispute(identifier, proposalTime);
+
+        // Get the returned funding rate from the oracle. If this has not yet resolved will revert.
+        // If the fundingRateDispute struct has been deleted, then this call will also fail because the proposal
+        // time will be 0.
+        FixedPoint.Signed memory settlementRate = _getOraclePrice(identifier, fundingRateDispute.proposal.time);
+
+        // Dispute was successful if settled rate is different from proposed rate.
+        bool disputeSucceeded = !settlementRate.isEqual(fundingRateDispute.proposal.rate);
+        address proposer = disputeSucceeded
+            ? fundingRateDispute.proposal.disputer
+            : fundingRateDispute.proposal.proposer;
+
+        if (disputeSucceeded) {
+            // If dispute succeeds:
+            // - pay disputer the dispute reward + final fee rebate
+
+            collateralCurrency.safeTransfer(
+                fundingRateDispute.proposal.disputer,
+                fundingRateDispute.proposal.finalFee.rawValue
+            );
+        } else {
+            // If dispute fails:
+            // - pay proposer the proposal reward + final fee rebate
+
+            collateralCurrency.safeTransfer(
+                fundingRateDispute.proposal.proposer,
+                fundingRateDispute.proposal.finalFee.rawValue
+            );
+        }
+
+        // Update current rate to settlement rate if there has not been a published funding rate since the dispute
+        // began.
+        if (_getLastPublishTimeForIdentifier(identifier) <= fundingRateDispute.proposal.time) {
+            FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+            uint256 currentTime = getCurrentTime();
+            fundingRateRecord.rate = settlementRate;
+            fundingRateRecord.publishTime = currentTime;
```

Comment:
Just curious -- why do we need publiishTime?

Reply:
So we can check if there has been a newer update since the dispute started. If so, then the resolved price isn't published as the current rate

Reply:
Hmmm, but then it seems like we may be comparing apples and oranges, right?

Are we comparing the _proposal time_ to the _publish time_? Shouldn't we either compare proposal time to proposal time or publish time to publish time?

Reply:
I would think just comparing the _proposal time_ would make the most sense.

Reply:
Yeah that's more reasonable I agree

Reply:
But we still need a way to determine the time that the **current proposal** was proposed. So I'm thinking we just change `publishTime`'s name to `proposeTime` in the `FundingRateRecord` struct

Reply:
Good point. Yeah, that makes sense to me.

Reply:
Implemented in this [commit](https://github.com/UMAprotocol/protocol/pull/2147/commits/be1ae8b2dd53363033f7eb799703a41dd87e0df9)